### PR TITLE
feat: Add Travelling Salesman Problem(Naive Approach)

### PR DESCRIPTION
**Pull request by Mayank17M**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1590"></a>

### CODE REVIEW

**Review by ayaankhan98 — CHANGES_REQUESTED** [review]
Hey, Thanks for your contribution. We request you to follow [contribution guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md)

**Comment by Mayank17M** [comment]
Made some changes. 
@ayaankhan98 Please check if it is appropriate now. Thanks!

File: graph/travelling_salesman_problem.cpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
+// Travelling salesman problem states that:
+// Given a list of cities and the distances between each pair of cities, what is
+// the shortest possible route that visits each city exactly once and returns to
+// the origin city?
+// This is the naive approach to solve this problem.
```
Comment (ayaankhan98) [inline graph/travelling_salesman_problem.cpp:5]: use doxygen documentation guidelines.

File: graph/travelling_salesman_problem.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+// References:
+// https://en.wikipedia.org/wiki/Travelling_salesman_problem
```
Comment (ayaankhan98) [inline graph/travelling_salesman_problem.cpp:8]: you can use markdown format to reference links

File: graph/travelling_salesman_problem.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+#include <algorithm>
+#include <iostream>
+#include <vector>
+using namespace std;
```
Comment (ayaankhan98) [inline graph/travelling_salesman_problem.cpp:13]: It's not a good practice to use `using namespace std;`

File: graph/travelling_salesman_problem.cpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    return min_path;
+}
+
+int main() {
+    vector<vector<int>> cities = {
+        {0, 20, 42, 35}, {20, 0, 30, 34}, {42, 30, 0, 12}, {35, 34, 12, 0}};
+    int V = cities.size();
+    cout << TravellingSalesmanProblem(cities, 0, V);
+}
```
Comment (ayaankhan98) [inline graph/travelling_salesman_problem.cpp:50]: Tests are missing.

**Review by ayaankhan98 — CHANGES_REQUESTED** [review]
Please follow [contribution guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md)

Thanks.

File: graph/travelling_salesman_problem.cpp [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+// References:
+// https://en.wikipedia.org/wiki/Travelling_salesman_problem
+
+#include <algorithm>
+#include <iostream>
+#include <vector>
+using namespace std;
+
+int TravellingSalesmanProblem(vector<vector<int>> cities, int src, int V) {
```
Comment (ayaankhan98) [inline graph/travelling_salesman_problem.cpp:15]: use reference to pass large structures like vectors or 2-D vectors.

Comment (ayaankhan98) [reply graph/travelling_salesman_problem.cpp:15]: or const references as per the need.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Thanks for reviewing, @ayaankhan98! 😄
@Mayank17M, in case you are confused or don't know what to do, you can check the [typical structure of a program](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program). I hope that helps. Let us know if you need any more help. 🙂

**Comment by Mayank17M** [comment]
Thankyou @Panquesito7 , I'll make the necessary changes.
Thanks! :)

File: graph/travelling_salesman_problem.cpp [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+ * (https://en.wikipedia.org/wiki/Travelling_salesman_problem)
+ *
+ * @author [Mayank Mamgain](http://github.com/Mayank17M)
+ *
+ * @details
+ * Travelling salesman problem asks:
+ * Given a list of cities and the distances between each pair of cities, what is
+ * the shortest possible route that visits each city exactly once and returns to
+ * the origin city?
+ * TSP can be modelled as an undirected weighted graph, such that cities are the
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:13]:
```suggestion
 * TSP can be modeled as an undirected weighted graph, such that cities are the
```

File: graph/travelling_salesman_problem.cpp [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+ * graph's vertices, paths are the graph's edges, and a path's distance is the
+ * edge's weight. It is a minimization problem starting and finishing at a
+ * specified vertex after having visited each other vertex exactly once.
+ * This is the naive implementation of the problem.
+ */
+
+#include <limits.h>
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:20]: Please use `limits` or `climits` instead.

File: graph/travelling_salesman_problem.cpp [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <vector>
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:25]: Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: graph/travelling_salesman_problem.cpp [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+/**
+ * @namespace graph
+ * @brief Graph Algorithms
+ */
+
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:31]:
```suggestion
```

File: graph/travelling_salesman_problem.cpp [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        min_path = std::min(min_path, curr_weight);
+
+    } while (next_permutation(vtx.begin(), vtx.end()));
+
+    return min_path;
+}
+}  // namespace graph
+
+/** Function to test the Algorithm */
+void tests() {
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:78]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
static void tests() {
```

File: graph/travelling_salesman_problem.cpp [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+/** Main function */
+int main() {
+    tests();
+    std::vector<std::vector<int>> cities = {
+        {0, 5, 10, 15}, {5, 0, 20, 30}, {10, 20, 0, 35}, {15, 30, 35, 0}};
+    int V = cities.size();
+    std::cout << graph::TravellingSalesmanProblem(&cities, 0, V);
+}
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:108]:
```suggestion
}

```

File: graph/travelling_salesman_problem.cpp [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+    std::cout << "Initiating Test 3..." << std::endl;
+    cities = {
+        {0, 10, 15, 20}, {10, 0, 35, 25}, {15, 35, 0, 30}, {20, 25, 30, 0}};
+    V = cities.size();
+    assert(graph::TravellingSalesmanProblem(&cities, 0, V) == 80);
+    std::cout << "Test 3 Passed..." << std::endl;
+}
+
+/** Main function */
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:101]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

File: graph/travelling_salesman_problem.cpp [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+int main() {
+    tests();
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:103]:
```suggestion
    tests();  // run self-test implementations
```

File: graph/travelling_salesman_problem.cpp [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    //// vtx stores the vertexs of the graph
+    std::vector<int> vtx;
+    for (int i = 0; i < V; i++) {
+        if (i != src) {
+            vtx.push_back(i);
+        }
+    }
+
+    //// store minimum weight Hamiltonian Cycle.
+    int min_path = INT_MAX;
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:55]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄👍

**Comment by Mayank17M** [comment]
I think all the requested changes have been covered. 
Please review @Panquesito7

File: graph/travelling_salesman_problem.cpp [hunk 15]
```diff
@@ -0,0 +1,114 @@
+/**
+ * @file
+ * @brief [Travelling Salesman Problem]
+ * (https://en.wikipedia.org/wiki/Travelling_salesman_problem)
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:4]:
```suggestion
 * (https://en.wikipedia.org/wiki/Travelling_salesman_problem) implementation
```

File: graph/travelling_salesman_problem.cpp [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+ * the origin city?
+ * TSP can be modeled as an undirected weighted graph, such that cities are the
+ * graph's vertices, paths are the graph's edges, and a path's distance is the
+ * edge's weight. It is a minimization problem starting and finishing at a
+ * specified vertex after having visited each other vertex exactly once.
+ * This is the naive implementation of the problem.
+ */
+
+#include <limits.h>  /// header for limits of integral types
+
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:21]:
```suggestion
#include <limits>  /// for limits of integral types
```

File: graph/travelling_salesman_problem.cpp [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+/**
+ * @brief Function calculates the minimum path distance that will cover all the
+ * cities starting from the source.
+ *
+ * @param cities matrix representation of cities
+ * @param src Point from where salesman is starting
+ * @param V number of vertices in the graph
+ *
+ */
+
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:42]:
```suggestion
```

File: graph/travelling_salesman_problem.cpp [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+        if (i != src) {
+            vtx.push_back(i);
+        }
+    }
+
+    //// store minimum weight Hamiltonian Cycle.
+    int min_path = INT_MAX;
+    do {
+        //// store current Path weight(cost)
+        int curr_weight = 0;
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:57]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). **Let me know if you need any help with that**.

File: graph/travelling_salesman_problem.cpp [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+ * @returns void
+ */
+static void tests() {
+    std::cout << "Initiatinig Predefined Tests..." << std::endl;
+    std::cout << "Initiating Test 1..." << std::endl;
+    std::vector<std::vector<int>> cities = {
+        {0, 20, 42, 35}, {20, 0, 30, 34}, {42, 30, 0, 12}, {35, 34, 12, 0}};
+    int V = cities.size();
+    assert(graph::TravellingSalesmanProblem(&cities, 0, V) == 97);
+    std::cout << "Test 1 Passed..." << std::endl;
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:87]:
```suggestion
    std::cout << "1st test passed..." << std::endl;
```

File: graph/travelling_salesman_problem.cpp [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+
+    std::cout << "Initiating Test 2..." << std::endl;
+    cities = {{0, 5, 10, 15}, {5, 0, 20, 30}, {10, 20, 0, 35}, {15, 30, 35, 0}};
+    V = cities.size();
+    assert(graph::TravellingSalesmanProblem(&cities, 0, V) == 75);
+    std::cout << "Test 2 Passed..." << std::endl;
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:93]:
```suggestion
    std::cout << "2nd test passed..." << std::endl;
```

File: graph/travelling_salesman_problem.cpp [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+
+    std::cout << "Initiating Test 3..." << std::endl;
+    cities = {
+        {0, 10, 15, 20}, {10, 0, 35, 25}, {15, 35, 0, 30}, {20, 25, 30, 0}};
+    V = cities.size();
+    assert(graph::TravellingSalesmanProblem(&cities, 0, V) == 80);
+    std::cout << "Test 3 Passed..." << std::endl;
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:100]:
```suggestion
    std::cout << "3rd test passed..." << std::endl;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! Great work! 😄👍
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/Mayank17M/C-Plus-Plus/actions

File: graph/travelling_salesman_problem.cpp [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    tests();  // run self-test implementations
+    std::vector<std::vector<int>> cities = {
+        {0, 5, 10, 15}, {5, 0, 20, 30}, {10, 20, 0, 35}, {15, 30, 35, 0}};
+    int V = cities.size();
+    std::cout << graph::TravellingSalesmanProblem(&cities, 0, V);
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:111]:
```suggestion
    std::cout << graph::TravellingSalesmanProblem(&cities, 0, V) << std::endl;
```

File: graph/travelling_salesman_problem.cpp [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+ * edge's weight. It is a minimization problem starting and finishing at a
+ * specified vertex after having visited each other vertex exactly once.
+ * This is the naive implementation of the problem.
+ */
+
+#include <algorithm>        /// for std::min
+#include <cassert>          /// for assert
+#include <iostream>         /// for IO operations
+#include <limits>           /// for limits of integral types
+#include <vector>           /// header for std::vector
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:24]:
```suggestion
#include <vector>           /// for std::vector
```

File: graph/travelling_salesman_problem.cpp [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+
+/**
+ * @namespace graph
+ * @brief Graph Algorithms
+ */
+namespace graph {
+
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:31]:
```suggestion
```

File: graph/travelling_salesman_problem.cpp [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+ * @param cities matrix representation of cities
+ * @param src Point from where salesman is starting
+ * @param V number of vertices in the graph
+ *
+ */
+int TravellingSalesmanProblem(std::vector<std::vector<int>> *cities, int src,
+                              int V) {
+    //// vtx stores the vertexs of the graph
+    std::vector<int> vtx;
+    for (int i = 0; i < V; i++) {
```
Comment (Panquesito7) [inline graph/travelling_salesman_problem.cpp:45]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). **Let me know if you need any help**. 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
This is all that needs to be addressed, and your PR will be merged! 🎉

**Comment by Mayank17M** [comment]
Addressed all the requested changes. Please review. Thanks!

**Review by Panquesito7 — APPROVED** [review]
Awesome work, @Mayank17M! It's been great for your first contribution here! We hope you keep contributing! 😄👍🎉

**Comment by amino19** [comment]
@Panquesito7: Here's the output, goes in:

image

Is it shows correct?

**Comment by Panquesito7** [comment]
> @Panquesito7: Here's the output, goes in:
> 
> image
> 
> Is it shows correct?

That looks correct for me. @Mayank17M, does that look correct?

**Comment by Mayank17M** [comment]
Yep, that's correct :)

**Review by amino19 — APPROVED** [review]
@Mayank17M: LGTM! :partying_face:

**Comment by Panquesito7** [comment]
Thanks for reviewing @amino19! 😄

**Pull request merged**